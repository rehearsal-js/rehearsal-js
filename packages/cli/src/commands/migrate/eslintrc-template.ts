module.exports = {
  parser: '',
  plugins: [
    '@typescript-eslint',
    'import',
    'simple-import-sort',
    'prettier',
    'filenames',
    'eslint-plugin-tsdoc',
    '@linkedin/typescript',
  ],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:prettier/recommended',
  ],
  rules: {
    '@typescript-eslint/naming-convention': [
      'error',
      {
        selector: 'typeLike',
        format: ['PascalCase'],
      },
      {
        selector: 'function',
        format: ['camelCase', 'PascalCase'],
      },
      {
        selector: ['classProperty', 'method', 'accessor'],
        format: ['camelCase'],
      },
      {
        selector: 'variableLike',
        format: ['camelCase', 'UPPER_CASE', 'PascalCase'],
      },
      {
        selector: 'parameter',
        format: ['camelCase'],
        leadingUnderscore: 'allow',
      },
      {
        selector: ['enumMember'],
        format: ['PascalCase', 'UPPER_CASE'],
      },
      {
        selector: ['property'],
        format: ['UPPER_CASE'],
        modifiers: ['readonly', 'static'],
      },
    ],
    'no-shadow': 'off',
    '@typescript-eslint/no-shadow': 'error',
    'tsdoc/syntax': 'error',
    '@linkedin/typescript/consistent-alias-convention': 'error',
    '@typescript-eslint/no-unused-vars': 'error',

    '@typescript-eslint/no-parameter-properties': 'error',
    '@typescript-eslint/no-use-before-define': ['error', { functions: false }],
    '@typescript-eslint/explicit-function-return-type': [
      'error',
      {
        allowExpressions: true,
        allowHigherOrderFunctions: true,
        allowDirectConstAssertionInArrowFunctions: true,
      },
    ],
    'sort-imports': 'off',
    'node/no-missing-import': 'off',
    'import/no-extraneous-dependencies': 'error',
    'import/no-unassigned-import': 'error',
    'import/no-duplicates': 'error',
    'simple-import-sort/imports': 'error',
    'simple-import-sort/exports': 'error',
    'import/no-mutable-exports': 'error',
    'no-new-wrappers': 'error',
    'no-new-object': 'error',
  },
  overrides: [
    {
      files: ['*.ts', '*.tsx', '*.tsm', '*.tsc'],
      parser: '@typescript-eslint/parser',
      parserOptions: {
        ecmaVersion: 'latest',
        sourceType: 'module',
      },
      settings: {
        'import/external-module-folders': ['node_modules', 'node_modules/@types'],
        'import/parsers': {
          '@typescript-eslint/parser': ['.ts', '.tsx'],
        },
        'import/resolver': {
          node: {
            extensions: ['.ts', '.tsx', '.tsm', '.tsc', '.json'],
          },
          typescript: {
            alwaysTryTypes: true,
          },
        },
      },
      rules: {
        '@typescript-eslint/no-explicit-any': 'error',
      },
    },
  ],
};
