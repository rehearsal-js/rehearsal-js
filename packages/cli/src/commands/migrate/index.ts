#!/usr/bin/env node
import { Command } from 'commander';
import { Listr } from 'listr2';
import { createLogger, format, transports } from 'winston';

import { version } from '../../../package.json';
import { parseCommaSeparatedList, gitIsRepoDirty, resetFiles } from '../../utils';
import {
  initTask,
  depInstallTask,
  convertTask,
  tsConfigTask,
  lintConfigTask,
  createScriptsTask,
} from './tasks';

import type { MigrateCommandOptions } from '../../types';

export const migrateCommand = new Command();

process.on('SIGINT', () => {
  console.log('Received SIGINT. Press Control-D to exit.');
});

migrateCommand
  .name('migrate')
  .description('migrate a javascript project to typescript')
  .option('-p, --basePath <project base path>', 'base directory of your project', process.cwd())
  .option('-e, --entrypoint <entrypoint>', 'entrypoint filepath of your project')
  .option(
    '-f, --format <format>',
    'report format separated by comma, e.g. -f json,sarif,md,sonarqube',
    parseCommaSeparatedList,
    ['sarif']
  )
  .option('-o, --outputPath <outputPath>', 'reports output directory', '.rehearsal')
  .option('-u, --userConfig <custom json config for migrate command>', 'path to rehearsal config')
  .option('-i, --interactive', 'interactive mode')
  .option('-v, --verbose', 'print debugging logs')
  .option('-d, --dryRun', 'print files that will be attempted to migrate', false)
  .action(async (options: MigrateCommandOptions) => {
    await migrate(options);
  });

async function migrate(options: MigrateCommandOptions): Promise<void> {
  const loggerLevel = options.verbose ? 'debug' : 'info';
  const logger = createLogger({
    transports: [new transports.Console({ format: format.cli(), level: loggerLevel })],
  });

  console.log(`@rehearsal/migrate ${version.trim()}`);

  if (!options.dryRun) {
    const hasUncommittedFiles = await gitIsRepoDirty(options.basePath);
    if (hasUncommittedFiles) {
      logger.warn(
        'You have uncommitted files in your repo. Please commit or stash them as Rehearsal will reset your uncommitted changes.'
      );
      process.exit(0);
    }
  }

  const defaultListrOption = {
    concurrent: false,
    exitOnError: true,
  };

  const tasks = [
    await initTask(options),
    await depInstallTask(options),
    await tsConfigTask(options),
    await lintConfigTask(options),
    await createScriptsTask(options),
  ];

  try {
    if (options.interactive) {
      // For issue #549, have to use simple renderer for the interactive edit flow
      // previous ctx is needed for the isolated convertTask
      const ctx = await new Listr(tasks, defaultListrOption).run();
      await new Listr([await convertTask(options, logger, ctx)], {
        renderer: 'simple',
        ...defaultListrOption,
      }).run();
    } else {
      await new Listr([...tasks, await convertTask(options, logger)], defaultListrOption).run();
    }
  } catch (e) {
    await resetFiles();
    logger.error(`${e}`);
  }
}
