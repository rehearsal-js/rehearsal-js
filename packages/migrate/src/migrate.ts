import { resolve, join } from 'node:path';
import findup from 'findup-sync';
import { readTSConfig, readJSON } from '@rehearsal/utils';
import {
  PluginsRunner,
  RehearsalService,
  isGlintFile,
  isGlintProject,
  type GlintService,
  DummyPlugin,
} from '@rehearsal/service';
import debug from 'debug';
import ts from 'typescript';
import {
  DiagnosticFixPlugin,
  DiagnosticCommentPlugin,
  isPrettierUsedForFormatting,
  LintPlugin,
  PrettierPlugin,
  ReRehearsePlugin,
  ServiceInjectionsTransformPlugin,
  DiagnosticReportPlugin,
} from '@rehearsal/plugins';
import fastGlob from 'fast-glob';
import {
  getGlintFixPlugin,
  getGlintReportPlugin,
  getGlintCommentPlugin,
  createGlintService,
} from './glint-utils.js';
import { getExcludePatterns } from '@rehearsal/migration-graph-shared';
import type { Reporter } from '@rehearsal/reporter';
import type { CompilerOptions } from 'typescript';
import type { TSConfig } from '@rehearsal/utils';

export type MigrateInput = {
  rootPath: string;
  sourceFilesAbs: string[];
  reporter: Reporter; // Reporter
  task?: { output: string };
  ignore?: string[];
};

export type MigrateOutput = {
  rootPath: string;
  configFile: string;
  sourceFilesAbs: string[];
};

const DEBUG_CALLBACK = debug('rehearsal:migrate');
const { parseJsonConfigFileContent } = ts;

export async function* migrate(input: MigrateInput): AsyncGenerator<string> {
  const basePath = resolve(input.rootPath);
  // these will always be typescript files
  const sourceFilesAbs = input.sourceFilesAbs;
  const reporter = input.reporter;
  const ignore = input.ignore || [];
  const configName = 'tsconfig.json';
  // output is only for tests
  const listrTask = input.task || { output: '' };
  const commentTag = '@rehearsal';
  const useGlint = await isGlintProject(basePath);

  DEBUG_CALLBACK('migration started');
  DEBUG_CALLBACK(`Base path: ${basePath}`);
  DEBUG_CALLBACK(`sourceFiles: ${JSON.stringify(sourceFilesAbs)}`);

  // read the tsconfig.json
  const configFile = readTSConfig<TSConfig>(resolve(basePath, configName));

  if (!configFile) {
    const message = `Config file '${configName}' not found in '${basePath}'`;
    throw Error(message);
  }

  const tsConfigOptions = configFile.compilerOptions as CompilerOptions;

  DEBUG_CALLBACK(`tsconfig file: ${configFile}`);

  const { fileNames: someFiles } = parseJsonConfigFileContent(
    configFile,
    ts.sys,
    basePath,
    {},
    configName
  );

  // these should NOT all go into the report
  const fileNames = [...new Set([...someFiles, ...input.sourceFilesAbs])];
  const ignoredPaths = ignore
    .flatMap((glob) => {
      return fastGlob.sync(glob, { cwd: basePath, ignore: getExcludePatterns() });
    })
    .map((filePath) => join(basePath, filePath));

  // remove ignored paths from the list of files to migrate
  const filesToMigrate = fileNames.filter((filePath) => !ignoredPaths.includes(filePath));
  const servicesMap = await readServiceMap(basePath, '.rehearsal/services-map.json');

  DEBUG_CALLBACK(`fileNames: ${JSON.stringify(filesToMigrate)}`);

  const service = useGlint
    ? await createGlintService(basePath)
    : new RehearsalService(tsConfigOptions, filesToMigrate);

  function isGlintService(
    service: GlintService | RehearsalService,
    useGlint: boolean
  ): service is GlintService {
    return service && useGlint;
  }

  const runner = new PluginsRunner({ basePath, service, reporter })
    // Resetting
    .queue(ReRehearsePlugin, {
      commentTag,
    })
    // Fix errors
    .queue(ServiceInjectionsTransformPlugin, {
      servicesMap,
    })
    .queue(
      DiagnosticFixPlugin,
      {
        safeFixes: true,
        strictTyping: true,
      },
      (fileName: string) => !(isGlintService(service, useGlint) && isGlintFile(service, fileName))
    )
    .queue(
      useGlint ? await getGlintFixPlugin() : DummyPlugin,
      (fileName: string) => isGlintService(service, useGlint) && isGlintFile(service, fileName)
    )
    // Fix formatting
    .queue(PrettierPlugin, (fileName: string) => isPrettierUsedForFormatting(fileName))
    .queue(
      LintPlugin,
      {
        eslintOptions: { cwd: basePath, useEslintrc: true, fix: true },
        reportErrors: false,
      },
      (fileName: string) => !isPrettierUsedForFormatting(fileName)
    )
    // Add ts-expect-error comments and report those errors
    .queue(
      DiagnosticCommentPlugin,
      {
        addHints: true,
        commentTag,
      },
      (fileName: string) => !(isGlintService(service, useGlint) && isGlintFile(service, fileName))
    )
    .queue(
      useGlint ? await getGlintCommentPlugin() : DummyPlugin,
      {
        commentTag,
      },
      (fileName: string) => isGlintService(service, useGlint) && isGlintFile(service, fileName)
    )
    // Format previously added comments
    .queue(PrettierPlugin, (fileName: string) => isPrettierUsedForFormatting(fileName))
    .queue(
      LintPlugin,
      {
        eslintOptions: { cwd: basePath, useEslintrc: true, fix: true },
        reportErrors: false,
      },
      (fileName: string) => !isPrettierUsedForFormatting(fileName)
    )
    .queue(
      DiagnosticReportPlugin,
      {
        commentTag,
      },
      (fileName: string) => !(isGlintService(service, useGlint) && isGlintFile(service, fileName))
    )
    .queue(
      useGlint ? await getGlintReportPlugin() : DummyPlugin,
      {
        commentTag,
      },
      (fileName: string) => isGlintService(service, useGlint) && isGlintFile(service, fileName)
    )
    // Report linter issues
    .queue(LintPlugin, {
      eslintOptions: { cwd: basePath, useEslintrc: true, fix: false },
      reportErrors: true,
    });

  yield* runner.run(filesToMigrate, { log: (message: string) => (listrTask.output = message) });

  // save report after all yields
  reporter.saveCurrentRunToReport(basePath);
}

async function readServiceMap(
  basePath: string,
  mapFilePattern: string
): Promise<Map<string, string> | undefined> {
  try {
    const maybeMapFile = findup(mapFilePattern, { cwd: basePath });
    if (maybeMapFile) {
      const map = await readJSON(maybeMapFile);
      return new Map<string, string>(Object.entries(map as { [key: string]: string }));
    }
  } catch (e) {
    /* empty */
  }

  return undefined;
}
