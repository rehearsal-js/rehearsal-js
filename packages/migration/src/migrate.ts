import ts from 'typescript';
import winston from 'winston';

import { parse, resolve } from 'path';

import RehearsalService from './rehearsal-service';

import DiagnosticAutofixPlugin from './plugins/diagnostics-autofix.plugin';
import EmptyLinesPreservePlugin from './plugins/empty-lines-preserve.plugin';
import EmptyLinesRestorePlugin from './plugins/empty-lines-restore.plugin';
import LintPlugin from './plugins/lint.plugin';
import Reporter from './reporter/reporter';

export type MigrateInput = {
  basePath: string;
  configName?: string;
  reportName?: string;
  reporter?: Reporter;
  logger?: winston.Logger;
};

export type MigrateOutput = {
  basePath: string;
  configFile: string;
  reportFile: string;
  sourceFiles: string[];
};

/**
 * Provides semantic diagnostic information in @ts-ignore comments and in a JSON report
 */
export default async function migrate(input: MigrateInput): Promise<MigrateOutput> {
  const basePath = resolve(input.basePath);
  const configName = input.configName || 'tsconfig.json';
  const reportName = input.reportName || '.rehearsal-report.json';
  const reporter = input.reporter;
  const logger = input.logger;

  const plugins = [
    LintPlugin,
    EmptyLinesPreservePlugin,
    DiagnosticAutofixPlugin,
    EmptyLinesRestorePlugin,
    LintPlugin,
  ];

  logger?.info('Migration started.');
  logger?.info(`Base path: ${basePath}`);

  const configFile = ts.findConfigFile(basePath, ts.sys.fileExists, configName);

  if (!configFile) {
    const message = `Config file '${configName}' not found in '${basePath}'`;
    logger?.error(message);
    throw Error(message);
  }

  logger?.info(`Config file found: ${configFile}`);

  const { config } = ts.readConfigFile(configFile, ts.sys.readFile);
  const { options, fileNames } = ts.parseJsonConfigFileContent(config, ts.sys, basePath);

  const service = new RehearsalService(options, fileNames);

  for (const fileName of fileNames) {
    logger?.info(`Processing file: ${fileName}`);

    for (const pluginClass of plugins) {
      const plugin = new pluginClass(service, logger, reporter);
      const updatedText = await plugin.run({ fileName });
      service.setFileText(fileName, updatedText);
    }

    // Save file to the filesystem
    service.saveFile(fileName);
  }

  logger?.info(`Migration finished.`);

  const reportFile = resolve(parse(configFile).dir, reportName);
  reporter?.save(reportFile);
  logger?.info(`Report saved to ${reportFile}`);

  return {
    basePath,
    configFile,
    reportFile,
    sourceFiles: fileNames,
  };
}
