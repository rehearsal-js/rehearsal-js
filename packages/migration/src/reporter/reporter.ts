import fs from 'fs';
import ts from 'typescript';
import winston from 'winston';
import json from './formatters/json';

export type ReportDiagnosticInfo = {
  file: string;
  code: number;
  message: string;
  start: number;
  length: number;
  nodeKind?: string;
  nodeText?: string;
};

export type Report = {
  version: string;
  diagnostics: ReportDiagnosticInfo[];
};

/**
 * Representation of diagnostic and migration report.
 */
export default class Reporter {
  readonly basePath: string;

  private report: Report;
  private logger?: winston.Logger;

  constructor(basePath = '', logger?: winston.Logger) {
    this.basePath = basePath;
    this.logger = logger?.child({ service: 'rehearsal-reporter' });
    this.report = {
      version: ts.version,
      diagnostics: [],
    };
  }

  getFileNames(): string[] {
    const fileNames: string[] = [];
    for (const diagnostic of this.report.diagnostics) {
      if (!fileNames.includes(diagnostic.file)) {
        fileNames.push(diagnostic.file);
      }
    }

    return fileNames;
  }

  getDiagnostics(fileName: string): ReportDiagnosticInfo[] {
    return this.report.diagnostics.filter((diagnostic) => diagnostic.file === fileName);
  }

  /**
   * Appends am information about provided diagnostic and related node to the report
   */
  addDiagnostic(diagnostic: ts.DiagnosticWithLocation, node?: ts.Node): void {
    this.report.diagnostics.push({
      file: diagnostic.file.fileName.replace(this.basePath, ''),
      code: diagnostic.code,
      message: ts.flattenDiagnosticMessageText(diagnostic.messageText, '. '),
      start: diagnostic.start,
      length: diagnostic.length,
      nodeKind: node ? ts.SyntaxKind[node.kind] : undefined,
      nodeText: node?.getText(),
    });
  }

  /**
   * Prints the current report using provided formatter (ex. json, pull-request etc.)
   */
  print(file: string, formatter: (report: Report) => string): string {
    const report = formatter(this.report);

    if (file) {
      fs.writeFileSync(file, report);
    }

    return report;
  }

  /**
   * Saves the current report information to the file in JSON format
   */
  save(file: string): void {
    this.print(file, json);
    this.logger?.info(`Report saved to: ${file}.`);
  }

  load(file: string): Reporter {
    if (!fs.existsSync(file)) {
      this.logger?.error(`Report file not found: ${file}.`);
    }

    this.logger?.info(`Report file found: ${file}.`);
    const content = fs.readFileSync(file, 'utf-8');
    const report = JSON.parse(content);

    if (!Reporter.isReport(report)) {
      this.logger?.error(`Report not loaded: wrong file format`);
    }

    this.report = report;
    this.logger?.info(`Report loaded from file.`);

    return this;
  }

  private static isReport(report: any): report is Report {
    return report && report.version !== undefined && report.diagnostics !== undefined;
  }
}
