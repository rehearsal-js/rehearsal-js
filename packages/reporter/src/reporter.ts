import { existsSync, readFileSync, writeFileSync } from 'fs';
import { DiagnosticCategory, flattenDiagnosticMessageText, SyntaxKind } from 'typescript';
import {
  type ProcessedFile,
  type Report,
  type ReportFormatter,
  type ReportItem,
  type Location,
} from './types';
import type { DiagnosticWithLocation, Node } from 'typescript';
import type { Logger } from 'winston';

type ReporterMeta = {
  projectName: string;
  basePath: string;
  commandName: string;
  tsVersion: string;
};

/**
 * Representation of diagnostic and migration report.
 */
export class Reporter {
  readonly basePath: string;

  public report: Report;
  private logger?: Logger;

  constructor(meta: ReporterMeta, logger?: Logger) {
    const { projectName, basePath, commandName, tsVersion } = meta;

    this.basePath = basePath;
    this.logger = logger?.child({ service: 'rehearsal-reporter' });

    this.report = {
      summary: {
        projectName: projectName,
        tsVersion: tsVersion,
        timestamp: new Date().toLocaleString('en-US', {
          year: 'numeric',
          month: 'numeric',
          day: 'numeric',
          hourCycle: 'h24',
          hour: '2-digit',
          minute: '2-digit',
          second: '2-digit',
        }),
        basePath: basePath,
        commandName: commandName,
      },
      items: [],
    };
  }

  getFileNames(): string[] {
    return [...new Set(this.report.items.map((item) => item.analysisTarget))];
  }

  getItemsByAnalysisTarget(fileName: string): ReportItem[] {
    return this.report.items.filter((item) => item.analysisTarget === fileName);
  }

  /**
   * Appends an element to the summary
   */
  addSummary(key: string, value: unknown): void {
    this.report.summary[key] = value;
  }

  /**
   * Appends am information about provided diagnostic and related node to the report
   */
  addItem(
    diagnostic: DiagnosticWithLocation,
    files: { [fileName: string]: ProcessedFile },
    fixed: boolean,
    node?: Node,
    triggeringLocation?: Location,
    hint = '',
    helpUrl = ''
  ): void {
    this.report.items.push({
      analysisTarget: diagnostic.file.fileName,
      files,
      errorCode: diagnostic.code,
      category: DiagnosticCategory[diagnostic.category],
      message: flattenDiagnosticMessageText(diagnostic.messageText, '. '),
      hint: hint,
      fixed,
      nodeKind: node ? SyntaxKind[node.kind] : undefined,
      nodeText: node?.getText(),
      helpUrl,
      //Original node that triggers the error; may not be the node where the codefix happens
      //Location in type ProcessedFile is where the codefix happens or the message is added
      //The two locations in most cases will be the same, but can be different
      nodeLocation: triggeringLocation || undefined,
    });
  }

  /**
   * Prints the current report using provided formatter (ex. json, pull-request etc.)
   */
  print(file: string, formatter: ReportFormatter): string {
    const report = formatter(this.report);

    if (file) {
      writeFileSync(file, report);
    }

    return report;
  }

  /**
   * Saves the current report information to the file in simple JSON format
   * to be able to load it later with 'load' function
   */
  save(file: string): void {
    this.print(file, (report: Report): string => JSON.stringify(report, null, 2));
    this.logger?.info(`Report saved to: ${file}.`);
  }

  /**
   * Loads the report exported by function 'save' from the file
   */
  load(file: string): Reporter {
    if (!existsSync(file)) {
      this.logger?.error(`Report file not found: ${file}.`);
    }

    this.logger?.info(`Report file found: ${file}.`);
    const content = readFileSync(file, 'utf-8');
    const report: Report = JSON.parse(content);

    if (!Reporter.isReport(report)) {
      this.logger?.error(`Report not loaded: wrong file format`);
      return this;
    }

    this.report = report as Report;
    this.logger?.info(`Report loaded from file.`);

    return this;
  }

  private static isReport(report: Report): report is Report {
    return report && report.summary !== undefined && report.items !== undefined;
  }
}
