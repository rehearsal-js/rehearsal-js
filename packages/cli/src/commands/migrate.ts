#!/usr/bin/env node

// TODO: handle ctrl + c

import { migrate } from '@rehearsal/migrate';
import {
  discoverEmberPackages,
  getMigrationStrategy,
  SourceFile,
} from '@rehearsal/migration-graph';
import { jsonFormatter, mdFormatter, Reporter, sarifFormatter } from '@rehearsal/reporter';
import { Command } from 'commander';
import { existsSync, readJSONSync, rmSync, writeJsonSync } from 'fs-extra';
import { Listr } from 'listr2';
import { resolve } from 'path';
import { createLogger, format, transports } from 'winston';

import { generateReports } from '../helpers/report';
import {
  MigrateCommandContext,
  MigrateCommandOptions,
  PackageMap,
  PackageSelection,
} from '../types';
import { UserConfig } from '../userConfig';
import { addDep, determineProjectName, parseCommaSeparatedList, runModuleCommand } from '../utils';

function ifHasTypescriptInDevdep(basePath: string): boolean {
  const packageJSONPath = resolve(basePath, 'package.json');
  const packageJSON = readJSONSync(packageJSONPath);
  return (
    (packageJSON.devDependencies && packageJSON.devDependencies.typescript) ||
    (packageJSON.dependencies && packageJSON.dependencies.typescript)
  );
}

export const migrateCommand = new Command();

migrateCommand
  .name('migrate')
  .description('Migrate Javascript project to Typescript')
  .requiredOption('-p, --basePath <project base path>', 'Base dir path of your project.')
  .option('-e, --entrypoint <entrypoint>', 'entrypoint js file for your project')
  .option(
    '-r, --report <reportTypes>',
    'Report types separated by comma, e.g. -r json,sarif,md',
    parseCommaSeparatedList,
    []
  )
  .option('-o, --outputPath <outputPath>', 'Reports output path', '.rehearsal')
  .option('-s, --strict', 'Use strict tsconfig file')
  .option('-c, --clean', 'Clean up old JS files after TS conversion')
  .option(
    '-u, --userConfig <custom json config for migrate command>',
    'File path for custom config'
  )
  .option('-i, --interactive', 'Interactive mode to help migrate part of large apps')
  .option('-v, --verbose', 'Print more logs to debug.')
  .action(async (options: MigrateCommandOptions) => {
    const loggerLevel = options.verbose ? 'debug' : 'info';
    const logger = createLogger({
      transports: [new transports.Console({ format: format.cli(), level: loggerLevel })],
    });

    const tasks = new Listr<MigrateCommandContext>(
      [
        {
          title: 'Initialization',
          task: async (_ctx, task) => {
            // _ctx.skip = true;
            // get custom config
            const userConfig = options.userConfig
              ? new UserConfig(options.basePath, options.userConfig, 'migrate')
              : undefined;

            _ctx.userConfig = userConfig;

            const projectName = await determineProjectName(options.basePath);

            if (options.interactive) {
              const packageSelections: PackageSelection[] = discoverEmberPackages(
                options.basePath
              ).map((p) => {
                let { packageName: name } = p;
                if (name === projectName) {
                  // a little better message to indicate this option is for entire project
                  name = `${name} (this entire project)`;
                }
                return {
                  name,
                  location: p.path,
                };
              });
              const packageMap = packageSelections.reduce((map, p) => {
                map[p.name] = p.location;
                return map;
              }, {} as PackageMap);

              _ctx.input = await task.prompt<{ test: boolean; other: boolean }>([
                {
                  type: 'Select',
                  name: 'packageSelection',
                  message:
                    'We have found multiple packages in your project, select the one to migrate:',
                  choices: Object.keys(packageMap),
                },
              ]);
              // update basePath based on the selection
              _ctx.targetPacakgePath = packageMap[_ctx.input as string];
              task.title = `Initializatoin Completed! Running migration on ${_ctx.input}.`;
            } else {
              _ctx.targetPacakgePath = options.basePath;
              task.title = `Initializatoin Completed! Running migration on ${projectName}.`;
            }

            // construct migration strategy and prepare all the files needs to be migrated
            const strategy = getMigrationStrategy(_ctx.targetPacakgePath, {
              entrypoint: options.entrypoint,
              filterByPackageName: [],
            });
            const files: Array<SourceFile> = strategy.getMigrationOrder();
            _ctx.strategy = strategy;
            _ctx.sourceFilesWithAbsolutePath = files.map((f) => f.path);
            _ctx.sourceFilesWithRelativePath = files.map((f) => f.relativePath);
          },
        },
        {
          title: 'Installing dependencies',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            // install custom dependencies
            if (_ctx.userConfig?.hasDependencies) {
              task.title = `Installing custom dependencies`;
              await _ctx.userConfig.install();
            }

            if (ifHasTypescriptInDevdep(options.basePath)) {
              task.skip('typescript already exists. Skipping installing typescript.');
            } else {
              await addDep(['typescript'], true, { cwd: options.basePath });
            }
          },
        },
        {
          title: 'Creating tsconfig.json',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            if (_ctx.userConfig?.hasTsSetup) {
              task.title = `Creating tsconfig from custom config.`;
              await _ctx.userConfig.tsSetup();
            } else {
              const configPath = resolve(options.basePath, 'tsconfig.json');

              if (existsSync(configPath)) {
                task.skip(`${configPath} already exists, skipping creating tsconfig.json`);
              } else {
                task.title = `Creating ${options.strict ? 'strict' : 'basic'} tsconfig.`;

                createTSConfig(
                  options.basePath,
                  _ctx.sourceFilesWithRelativePath,
                  !!options.strict
                );
              }
            }
          },
        },
        {
          title: 'Converting JS files to TS',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            const projectName = (await determineProjectName()) || '';
            const reporter = new Reporter(projectName, options.basePath, logger);

            if (_ctx.sourceFilesWithAbsolutePath) {
              const input = {
                basePath: _ctx.targetPacakgePath,
                sourceFiles: _ctx.sourceFilesWithAbsolutePath,
                logger: options.verbose ? logger : undefined,
                reporter,
              };

              await migrate(input);

              const reportOutputPath = resolve(options.basePath, options.outputPath);
              generateReports(reporter, reportOutputPath, options.report, {
                json: jsonFormatter,
                sarif: sarifFormatter,
                md: mdFormatter,
              });
            } else {
              task.skip(
                `Skipping JS -> TS conversion task, since there is no JS file to be converted to TS.`
              );
            }
          },
        },
        {
          title: 'Clean up old JS files',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            if (_ctx.sourceFilesWithAbsolutePath.length && options.clean) {
              cleanJSFiles(_ctx.sourceFilesWithAbsolutePath);
            } else {
              task.skip(`Skipping clean up task`);
            }
          },
        },
        {
          title: 'Run Typescript compiler to check errors',
          enabled: (ctx): boolean => !ctx.skip,
          task: async () => {
            await runModuleCommand(['tsc'], { cwd: options.basePath });
          },
        },
        // TODO: what to do with those ts errors?

        {
          title: 'Creating lint config',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            if (_ctx.userConfig?.hasLintSetup) {
              task.title = `Creating .eslintrc.js from custom config.`;
              await _ctx.userConfig.lintSetup();
            } else {
              task.skip(`Skip creating .eslintrc.js since no custom config is provided.`);
            }
          },
        },
      ],
      { concurrent: false, exitOnError: false }
    );
    try {
      await tasks.run();
    } catch (e) {
      logger.error(`${e}`);
    }
  });

/**
 * Generate tsconfig
 */
function createTSConfig(basePath: string, fileList: string[], strict: boolean): void {
  const include = [...fileList.map((f) => f.replace('.js', '.ts'))];
  const config = strict
    ? {
        $schema: 'http://json.schemastore.org/tsconfig',
        compilerOptions: {
          baseUrl: '.',
          outDir: 'dist',
          strict: true,
          noUncheckedIndexedAccess: true,
          module: 'es2020',
          moduleResolution: 'node',
          newLine: 'lf',
          forceConsistentCasingInFileNames: true,
          noFallthroughCasesInSwitch: true,
          target: 'es2020',
          lib: ['es2020'],
          esModuleInterop: true,
          declaration: true,
          sourceMap: true,
          declarationMap: true,
        },
        include,
      }
    : {
        $schema: 'http://json.schemastore.org/tsconfig',
        compilerOptions: {
          baseUrl: '.',
          outDir: 'dist',
          emitDeclarationOnly: true,
          allowJs: true,
          target: 'es2016',
          module: 'commonjs',
          moduleResolution: 'node',
        },
        include,
      };
  writeJsonSync(resolve(basePath, 'tsconfig.json'), config, { spaces: 2 });
}

/**
 * Remove files
 * @param fileList Array of file paths
 */
function cleanJSFiles(fileList: string[]): void {
  fileList.filter((f) => f.match(/\.js$/)).forEach((f) => rmSync(f));
}
