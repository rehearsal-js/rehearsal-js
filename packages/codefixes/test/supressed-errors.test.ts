import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, expect, test } from 'vitest';
import { Project } from 'fixturify-project';
import { Reporter } from '@rehearsal/reporter';
import { migrate } from '@rehearsal/migrate';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe('Supressed Errors', function () {
  const cases = [
    [`TS2345: Cannot read properties of undefined (reading 'flags')`, 'supressed-errors/flags.ts'],
    ['TS2339: False expression: Token end is child end', 'supressed-errors/missingMethod.ts'],
  ];

  test.each(cases)('%s, %s', async (name, relativePathToFixture) => {
    const fixtureDir = resolve(__dirname, 'fixtures');
    const project = Project.fromDir(fixtureDir, { linkDeps: true, linkDevDeps: true });
    await project.write();

    const file = resolve(project.baseDir, relativePathToFixture);

    const reporter = new Reporter({
      tsVersion: '',
      projectName: '@rehearsal/test',
      projectRootDir: project.baseDir,
      commandName: '@rehearsal/migrate',
    });

    const input = {
      projectRootDir: project.baseDir,
      packageDir: project.baseDir,
      filesToMigrate: [file],
      reporter,
    };

    for await (const _ of migrate(input)) {
      // no ops
    }

    const actualOutput = readFileSync(file, 'utf-8');

    expect(actualOutput).matchSnapshot();

    project.dispose();
  });
});
