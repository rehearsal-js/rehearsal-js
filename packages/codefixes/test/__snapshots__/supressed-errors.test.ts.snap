// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Supressed Errors > TS2339: False expression: Token end is child end, supressed-errors/missingMethod.ts 1`] = `
"// @ts-expect-error @rehearsal TODO TS6133: The declaration 'Base' is never read or used. Remove the declaration or use it.
import { Base } from \\"some-package\\";
"
`;

exports[`Supressed Errors > TS2345: Cannot read properties of undefined (reading 'flags'), supressed-errors/flags.ts 1`] = `
"(function() {
  const config = {
    values: [],
    value: {},
  };

  // @ts-expect-error @rehearsal TODO TS2345: Argument of type '{0}' is not assignable to parameter of type 'never'. Consider verifying both types, using type assertion: '(config.value as string)', or using type guard: 'if (config.value instanceof string) { ... }'.
  config.values.push(config.value);
}();
"
`;
